### PR TITLE
Add StaticTest.vm

### PR DESCRIPTION
**Pull request by wildmaples**
The last `.vm` file to implement in our chapter.

### CODE REVIEW

File: test/fixtures/StaticTest.asm [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+A=M
+M=D
+@SP
+M=M+1
+@SP
+AM=M-1
+D=M
+@R13
+M=D
+@24
```
Comment (tomstuart) [inline test/fixtures/StaticTest.asm:27]:
```suggestion
@StaticTest.8
```

File: test/fixtures/StaticTest.asm [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+A=M
+M=D
+@SP
+M=M+1
+@SP
+AM=M-1
+D=M
+A=A-1
+M=M-D
+@24
```
Comment (tomstuart) [inline test/fixtures/StaticTest.asm:83]:
```suggestion
@StaticTest.8
```

File: test/fixtures/StaticTest.asm [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+D=M
+@R14
+A=M
+M=D
+@SP
+AM=M-1
+D=M
+@R13
+M=D
+@19
```
Comment (tomstuart) [inline test/fixtures/StaticTest.asm:41]:
```suggestion
@StaticTest.3
```

File: test/fixtures/StaticTest.asm [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+@17
+D=A
+@R14
+M=D
+@R13
+D=M
+@R14
+A=M
+M=D
+@19
```
Comment (tomstuart) [inline test/fixtures/StaticTest.asm:64]:
```suggestion
@StaticTest.3
```

File: test/fixtures/StaticTest.asm [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+D=M
+@R14
+A=M
+M=D
+@SP
+AM=M-1
+D=M
+@R13
+M=D
+@17
```
Comment (tomstuart) [inline test/fixtures/StaticTest.asm:55]:
```suggestion
@StaticTest.1
```

File: test/fixtures/StaticTest.asm [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+A=M
+M=D
+@19
+D=M
+@SP
+A=M
+M=D
+@SP
+M=M+1
+@17
```
Comment (tomstuart) [inline test/fixtures/StaticTest.asm:71]:
```suggestion
@StaticTest.1
```

**Review by tomstuart — CHANGES_REQUESTED** [review]
Sorry, this isn’t right! Confusingly it does make the `StaticTest.tst` test pass, so I understand why you did it, but it doesn’t meet the specification of the `static` segment at the top of page 143:

> According to the Hack machine language specification, when a new symbol is encountered for the first time in an assembly program, the assembler allocates a new RAM address to it, starting at address 16. This convention can be exploited to represent each static variable number `j` in a VM file `f` as the assembly language symbol `f.j`. For example, suppose that the file `Xxx.vm` contains the command `push static 3`. This command can be translated to the Hack assembly commands `@Xxx.3` and `D=M`, followed by additional assembly code that pushes `D`’s value to the stack. This implementation of the `static` segment is somewhat tricky, but it works.

“Somewhat tricky” indeed. The idea is to depend upon the [existing behaviour](https://github.com/wildmaples/ruby-hack-assembler/blob/main/test/unit/assembler_test.rb#L59-L63) of the Hack assembler, which assigns user-defined “variable symbols” to distinct addresses in the symbol table, [starting at address 16](https://github.com/wildmaples/ruby-hack-assembler/blob/main/lib/assembler.rb#L11) and [auto-incrementing the address](https://github.com/wildmaples/ruby-hack-assembler/blob/main/lib/assembler.rb#L72-L73) each time a new symbol is added.

Concretely, the A-command that you generate from `pop static 8` to load the destination address into the A register should be `@StaticTest.8`, not `@24`, so that the assembler gets to decide which address to assign to the new `StaticTest.8` symbol. The reason `StaticTest.tst` passes regardless is that [`StaticTest.cmp`](https://github.com/wildmaples/nand2tetris/blob/master/projects/07/MemoryAccess/StaticTest/StaticTest.cmp) doesn’t actually check which addresses you used, only that the correct value ends up on the top of the stack. So the fact that you’re using addresses 24, 19 and 17 instead of letting the assembler decide where to store `@StaticTest.8`, `@StaticTest.3` and `@StaticTest.1` (i.e. addresses 16, 17 and 18) isn’t visible to the test.

So why does it even matter? Because, although we haven’t tried it yet, the VM translator has to support translating multiple files at once (e.g. for [`FibonacciElement`](https://github.com/wildmaples/nand2tetris/tree/master/projects/08/FunctionCalls/FibonacciElement) in the next chapter) into a single `.asm` file. In theory we already learned about this on page 143:

> The VM translator should accept a single command line parameter, as follows: 
>
> `prompt> VMtranslator source`
>
>
> Where _source_ is either a file name of the form `Xxx.vm` (the extension is mandatory) or a directory name containing one or more `.vm` files (in which case there is no extension). The result of the translation is always a single assembly language file named `Xxx.asm`, created in the same directory as the input `Xxx`. The translated code must conform to the standard VM mapping on the Hack platform. 

And on page 146:

> If the program’s argument is a directory name rather than a file name, the main program should process all the `.vm` files in this directory. In doing so, it should use a separate `Parser` for handling each input file and a single `CodeWriter` for handling the output. 

In practice we haven’t had to deal with any of this because we haven’t needed to translate multiple VM files yet. But when we do, it’s important that `Foo.vm` can read and write to `static 8` without interfering with `Bar.vm`’s `static 8`. If we always translate `static 8` into address 24 then those two VM files will be fighting over the use of the same memory location; the solution is to translate one into `@Foo.8` and the other into `@Bar.8`, so that the assembler will take care of assigning those variable symbols to distinct addresses (e.g. 16 and 17) that don’t clash.

File: bin/vm-translator [hunk 7]
```diff
@@ -2,5 +2,9 @@
 
 require_relative '../lib/vm_translator'
 
-input_file = File.open(ARGV[0].to_s)
-puts VMTranslator.new(input_file).translate
+if File.file?(ARGV[0])
+  input_file = File.open(ARGV[0])
+  puts VMTranslator.new(input_file).translate
+else
+  raise ArgumentError.new("We can only translate a single .vm file at a time")
```
Comment (wildmaples) [inline bin/vm-translator:9]: https://github.com/wildmaples/jack-virtual-machine/issues/15

**Review by tomstuart — APPROVED** [review]
This looks good, well done! I think we’ll want to revisit the `VMTranslator` setup once we need to translate multiple files at once, but we’re not there yet, so this is good right now. 🎉

**Pull request merged**